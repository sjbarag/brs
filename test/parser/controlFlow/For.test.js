const BrsError = require("../../../lib/Error");
const { Lexeme, BrsTypes, Parser } = require("brs");
const { Int32 } = BrsTypes;

const { EOF } = require("../ParserTests");

describe("parser for loops", () => {
    afterEach(() => BrsError.reset());

    it("accepts a 'step' clause", () => {
        let parsed = Parser.parse([
            { kind: Lexeme.For, text: "for", line: 1 },
            { kind: Lexeme.Identifier, text: "i", line: 1 },
            { kind: Lexeme.Equal, text: "=", line: 1 },
            { kind: Lexeme.Integer, text: "0", literal: new Int32(0), line: 1 },
            { kind: Lexeme.To, text: "to", line: 1 },
            { kind: Lexeme.Integer, text: "5", literal: new Int32(5), line: 1 },
            { kind: Lexeme.Step, text: "step", line: 1 },
            { kind: Lexeme.Integer, text: "2", literal: new Int32(2), line: 1 },
            { kind: Lexeme.Newline, text: "\n", line: 1 },
            // body would go here, but it's not necessary for this test
            { kind: Lexeme.EndFor, text: "end for", line: 2 },
            { kind: Lexeme.Newline, text: "\n", line: 2 },
            EOF
        ]);

        expect(BrsError.found()).toBe(false);
        expect(parsed).toBeDefined();
        expect(parsed[0]).toBeDefined();
        expect(parsed[0].increment).toBeDefined();
        expect(parsed[0].increment.value).toEqual(new Int32(2));

        expect(parsed).toMatchSnapshot();
    });

    it("defaults a missing 'step' clause to '1'", () => {
        let parsed = Parser.parse([
            { kind: Lexeme.For, text: "for", line: 1 },
            { kind: Lexeme.Identifier, text: "i", line: 1 },
            { kind: Lexeme.Equal, text: "=", line: 1 },
            { kind: Lexeme.Integer, text: "0", literal: new Int32(0), line: 1 },
            { kind: Lexeme.To, text: "to", line: 1 },
            { kind: Lexeme.Integer, text: "5", literal: new Int32(5), line: 1 },
            { kind: Lexeme.Newline, text: "\n", line: 1 },
            // body would go here, but it's not necessary for this test
            { kind: Lexeme.EndFor, text: "end for", line: 2 },
            { kind: Lexeme.Newline, text: "\n", line: 2 },
            EOF
        ]);

        expect(BrsError.found()).toBe(false);
        expect(parsed).toBeDefined();
        expect(parsed[0]).toBeDefined();
        expect(parsed[0].increment).toBeDefined();
        expect(parsed[0].increment.value).toEqual(new Int32(1));

        expect(parsed).toMatchSnapshot();
    });
});
