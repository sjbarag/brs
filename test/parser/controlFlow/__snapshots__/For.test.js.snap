// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser for loops accepts a 'step' clause 1`] = `
Array [
  For {
    "body": Block {
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "file": undefined,
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "statements": Array [],
      "type": "Block",
    },
    "counterDeclaration": Assignment {
      "name": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "i",
      },
      "tokens": Object {
        "equals": Object {
          "isReserved": false,
          "kind": "Equal",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "=",
        },
      },
      "type": "Assignment",
      "value": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": Int32 {
          "kind": 4,
          "value": 0,
        },
      },
    },
    "finalValue": Literal {
      "_location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "type": "Literal",
      "value": Int32 {
        "kind": 4,
        "value": 5,
      },
    },
    "increment": Literal {
      "_location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "type": "Literal",
      "value": Int32 {
        "kind": 4,
        "value": 2,
      },
    },
    "tokens": Object {
      "endFor": Object {
        "isReserved": false,
        "kind": "EndFor",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "end for",
      },
      "for": Object {
        "isReserved": true,
        "kind": "For",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "for",
      },
      "step": Object {
        "isReserved": true,
        "kind": "Step",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "step",
      },
      "to": Object {
        "isReserved": true,
        "kind": "To",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "to",
      },
    },
    "type": "For",
  },
]
`;

exports[`parser for loops allows 'next' to terminate loop 1`] = `
Array [
  For {
    "body": Block {
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "file": undefined,
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "statements": Array [],
      "type": "Block",
    },
    "counterDeclaration": Assignment {
      "name": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "i",
      },
      "tokens": Object {
        "equals": Object {
          "isReserved": false,
          "kind": "Equal",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "=",
        },
      },
      "type": "Assignment",
      "value": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": Int32 {
          "kind": 4,
          "value": 0,
        },
      },
    },
    "finalValue": Literal {
      "_location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "type": "Literal",
      "value": Int32 {
        "kind": 4,
        "value": 5,
      },
    },
    "increment": Literal {
      "_location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "type": "Literal",
      "value": Int32 {
        "kind": 4,
        "value": 1,
      },
    },
    "tokens": Object {
      "endFor": Object {
        "isReserved": true,
        "kind": "Next",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "next",
      },
      "for": Object {
        "isReserved": true,
        "kind": "For",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "for",
      },
      "step": undefined,
      "to": Object {
        "isReserved": true,
        "kind": "To",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "to",
      },
    },
    "type": "For",
  },
]
`;

exports[`parser for loops defaults a missing 'step' clause to '1' 1`] = `
Array [
  For {
    "body": Block {
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "file": undefined,
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "statements": Array [],
      "type": "Block",
    },
    "counterDeclaration": Assignment {
      "name": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "i",
      },
      "tokens": Object {
        "equals": Object {
          "isReserved": false,
          "kind": "Equal",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "=",
        },
      },
      "type": "Assignment",
      "value": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": Int32 {
          "kind": 4,
          "value": 0,
        },
      },
    },
    "finalValue": Literal {
      "_location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "type": "Literal",
      "value": Int32 {
        "kind": 4,
        "value": 5,
      },
    },
    "increment": Literal {
      "_location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "type": "Literal",
      "value": Int32 {
        "kind": 4,
        "value": 1,
      },
    },
    "tokens": Object {
      "endFor": Object {
        "isReserved": false,
        "kind": "EndFor",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "end for",
      },
      "for": Object {
        "isReserved": true,
        "kind": "For",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "for",
      },
      "step": undefined,
      "to": Object {
        "isReserved": true,
        "kind": "To",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "to",
      },
    },
    "type": "For",
  },
]
`;
